enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,423 @@
+  where the Route is going to be created.
+- To expose a user workload, the user would create a new Route with an annotation
+  referencing the generated secret that was created in the previous step.
+- If the secret that is referenced exists and has a successfully generated
+  cert/key pair generated, the secret-injector controller copies the certificate data
+  into the route at `.spec.tls.certificate` and `.spec.tls.key`.
+
+**When an Ingress is used to expose user workload**
+
+- The end user must have generated the serving certificate as a pre requisite
+  using third-party systems like cert-manager.
+- In cert-manager's case, the [Certificate](https://cert-manager.io/docs/usage/certificate/#creating-certificate-resources) CR must be created in the same namespace
+  where the Ingress is going to be created.
+- To expose a user workload, he would need to create a new Ingress with the generated secret
+  referenced in `.spec.tls.secretName`.
+- If the secret CR that is referenced exists and has a successfully generated
+  cert/key pair, the ingress-to-route controller copies this certificate data
+  into the route at `.spec.tls.certificate` and `.spec.tls.key`.
+
+_Note_: The same workflow would apply to updating an existing Route to integrate
+with cert-manager. The new certificate data will overwrite the `.spec.tls.certificate`
+and `.spec.tls.key` in the route CR since the annotation gets a higher preference.
+
+#### Certificate Life cycle
+
+Post integrating with third-party solutions like cert-manager, the end user can also
+disable automatic certificate management. This can be done in 2 ways,
+
+- The end user can delete the annotation on the Route.
+- The end user deletes the secret that was associated with the Certificate CR.
+
+**Deleting the annotation on the Route**
+The end user can delete the annotation on the Route, this will result
+in abondoning the cert/key pair that was generated via cert-manager and resort to
+using the default certificates that were initially generated by the cluster-ingress-operator
+i.e. clearing the `.spec.tls` on Route.
+
+**Deleting the generated secret that contains the serving cert/key pair**
+When the end user deletes the secret containing the certificate data, the `.spec.tls` on
+Route will be cleared. This will result in the route serving default certificates. In such
+a scenario the annotation that was added by the user will be automatically deleted by the
+secret-injector controller. This is only in the case where the user is directly creating
+routes instead of relying on the Ingresses.
+
+#### Variation [optional]
+
+N.A
+
+### API Extensions
+
+A new annotation of type `string` to be introduced as part of the integration
+to support third-party certificate management systems,
+
+```yaml
+annotations:
+  route.openshift.io/tls-secret-name: <secret-name>
```

Comment:
I haven't reviewed this enhancement in detail, but a quick look at the API extensions section, I would lean towards using a new API field over the annotation approach. Creating annotation APIs is not discoverable and does not allow for validation or versioning, they are effectively, free form and guess work.

The `CertificateReference` in the `RouteSpec` is the way our API conventions would suggest you go

Reply:
So in this context the alternative proposal of `CertificateReference` would essentially point to a third party CRD like [Certificate](https://cert-manager.io/docs/usage/certificate/#creating-certificate-resources) CRD provided by the `cert-manager`.

If what you are suggesting is that the `CertificateReference` (basically a `secretRef`/`secretName`) would just point to Secret of tls type, the route `.spec` would end up have multiple fields are related to the certificate data but they would be a oneoff (`.spec.tls.certificate`, `.spec.tls.key` OR `.spec.tls.certificateReference`). Is this what you are suggesting?

Reply:
The annotation was also kind of the preference of NE team initially. @alebedev87 and @Miciah Since we had put a pin on this until API review, do you think we can switch to a field (secretName) instead of an annotation?  

Another instance of this is seen in the gateway API as well as [CertificateRef](https://github.com/kubernetes-sigs/gateway-api/blob/df03bd58002fcf01ee3ffeba97edc1f0c67fb386/apis/v1beta1/gateway_types.go#L334) (although it is a cross namespace secret which we don't support)

Reply:
I believe there is another caveat to making this a field. Currently the router reads the certificate data from `.spec.tls.certificate` and `.spec.tls.key`, and it was a deliberate choice not read this directly from a secret (this reduces permissions required by the router and reduces attach surface on the edge). 

If we introduce a field, we still won't be able to remove `.spec.tls.certificate` and `.spec.tls.key`, so the logic would be something like, the secret is read and then copied onto the other 2 fields. Essentially one field helps in populating the other 2 field and this kind of seems weird (an annotation seems better suited to kind of configure this behaviour).

Reply:
> Is having secret data embedded in this object not a security risk in itself? At least with a secret object users can introduce KMS encryption to protect the data, they don't have that option with data stored in a CRD. 

I hadn't heard of KMS before now.  Is KMS encryption incompatible with aggregated API servers?  The KMS documentation specifically calls out CRDs as supported in Kubernetes 1.26, but the documentation doesn't mention aggregated API servers such as openshift-apiserver, which handles the Route API in OpenShift: https://kubernetes.io/docs/tasks/administer-cluster/kms-provider/#encrypting-your-data-with-the-kms-provider

(By the way, I've asked various people what makes secrets _actually_ more secret than other resources that are restricted by RBAC.  This might be the first substantive answer to that question that I've seen.  :tada:!)

> I would suggest revisiting that decision. Out of interest, were the pros and cons of reading secrets directly documented somewhere? Perhaps an arch discussion or enhancement can help give us the background?

The decision was made many years ago.  I don't know where it was written down, but it's basically this:  The openshift-router and haproxy processes run in the same pod, the haproxy process receives connections from untrusted clients and forwards connections to untrusted backend server pods, so it should be kept far away from anything particularly privileged, and so we don't want openshift-router to be privileged enough to read secrets.  

In short, the design that Thejas is proposing (and the design of the existing ingress-to-route controller) provides security through separation of concerns: one pod (with the secret-injector and ingress-to-route controllers) reads secrets and updates routes, and a different pod (the router) reads routes and handles network connections.  

The existence of KMS (which apparently can make some resources more secure than other resources) and the ability to do name-scoped watches on RBAC-restricted resources (which I learned about today) change the pro/con balance.  However, I'm wary of re-architecting openshift-router to establish and manage potentially hundreds of watches on secrets in the short time we have to implement this EP.

Reply:
I feel like I'm having parallel conversations about this, but don't see the other thread 🤔 

I suspect you're right that the KMS integration isn't going to work with an aggregated API server, IIUC the requests for aggregated APIs are only proxied by the primary API server, so it wouldn't transform and implement the KMS encryption. That is a definite advantage to using a secret over an aggregated API, but you're right, we could also KMS encrypt a CRD.

IMO the main thing that makes secrets special is education, it's what users expect to be the most secure, because that's what the kube community has told them for a long time. Secrets were the first to have at least some encryption before going into storage, we now hacked that for aggregated APIs and CRDs too. Secrets are the most obvious target for KMS, but you can add CRDs there too. Ask the average user what to encrypt and I bet the first thing they say is a Secret 😉 

> In short, the design that Thejas is proposing (and the design of the existing ingress-to-route controller) provides security through separation of concerns: one pod (with the secret-injector and ingress-to-route controllers) reads secrets and updates routes, and a different pod (the router) reads routes and handles network connections.

That's a fair argument and thanks for the history

> However, I'm wary of re-architecting openshift-router to establish and manage potentially hundreds of watches on secrets in the short time we have to implement this EP.

I'm curious what the time pressure is here? My take on what's happening in OpenShift right now is that we are supposed to be focusing on quality and that deadlines for features and hitting specific releases shouldn't be our primary concern, if we think there's a better way to do the feature or rushing it will cause issues down the line, we should be able to say no, this isn't going to make 4.13. If missing that deadline means we can't ship 4.13, that's probably a different conversation but I would encourage you to think about the balance of a (arbitrary?) deadline vs getting the feature right

Reply:
Hey @JoelSpeed,
There were parallel discussions on this and in light of the recent discussions and implementation suggestions, we've decided to pick this up for the next release and not for 4.13 GA. So i'm putting this on hold for now and will pick this later (post 4.13 GA).